### ISSUE TITLE
%typemap(in)      (char *STRING, int LENGTH) with types other than char*

### ISSUE THREAD

**Issue by stevenhdb** [swig/swig#248]
Hi,

I'm using SWIG 3.0.2 and came across an issue where I needed to use the above typemap on a BYTE*.

I added `%apply (char *STRING, int LENGTH) { (BYTE* data, int len) }` to my code but got a compile error in the generated wrapper cxx code on the line `arg2 = (char *) jenv->GetByteArrayElements(jarg2, 0);`

The fix for this seems to be to change the typemap in java.swg to use `$1_type` instead of char \* as follows:

```
@@ -1327,7 +1327,7 @@ SWIG_PROXY_CONSTRUCTOR(true, true, SWIGTYPE)
 %typemap(freearg) (char *STRING, size_t LENGTH) ""
 %typemap(in)      (char *STRING, size_t LENGTH) {
   if ($input) {
-    $1 = (char *) JCALL2(GetByteArrayElements, jenv, $input, 0);
+    $1 = ($1_type) JCALL2(GetByteArrayElements, jenv, $input, 0);
     $2 = (size_t) JCALL1(GetArrayLength, jenv, $input);
   } else {
     $1 = 0;
```

Is this a valid change?

Thanks,
Steven

**Comment by ojwb** [comment]
@wsfulton Probably a question for you...

**Comment by wsfulton** [comment]
Yes. Just needs applying with a testcase which will probably flush out similar problems in other languages.

**Comment by vadz** [comment]
See also this [ML thread](https://sourceforge.net/p/swig/mailman/message/34330407/)

**Comment by ojwb** [comment]
This was fixed back in 2015 by a1bddd56eb3404c9a069a48fa7b391fd255a7905 (in SWIG 3.0.8).
